components/spinner.py

```diff
@@ -3,29 +3,128 @@
+        "Y": Colour(72 / 255, 131 / 255, 50 / 255),
+    }
+
+    WHEEL_ORDER = ["R", "Y", "B", "G"]
+
+    POSITION_SPIN_FACTOR = 0.3  # position control will run at up to twice this speed
+
+    # rotation control uses the formula: 1 - ROTATION_SPIN_FACTOR ** (rotations - ROTATION_TARGET_ROTATIONS)  for wheel speed
+    ROTATION_SPIN_FACTOR = 20  # increase this to recude slow down
+    ROTATION_MAX_SPEED = 1
+    ROTATION_TARGET_ROTATIONS = 3.5
+
     def __init__(self):
         pass
 
     def setup(self):
-        pass
+        self.spinner_motor.set(0)
+        self.raise_wheel()
+        self.rotations = 0
+        # self.state = None
+        self.piston_state = "down"
+        self.required_colour = "B"
+        self.lastCol = "Y"
 
-    def on_enable(self):
-        self.spinner_motor.stopMotor()
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kReverse)
+    def on_enable(self) -> None:
+        self.spinner_motor.set(0)
+        self.raise_wheel()
 
-    def execute(self):
-        pass
+    def raise_wheel(self) -> None:  # moves wheel up
+        self.spinner_solenoid.set(False)
+        self.piston_state = "up"
 
-    def read_colour(self):
-        return (self.colour_sensor.getColor(), self.colour_sensor.getIR())
+    def lower_wheel(self) -> None:  # moves wheel down
+        self.spinner_solenoid.set(True)
+        self.piston_state = "down"
 
-    def piston_up(self):
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kForward)
+    def get_current_colour(self) -> str:  # reads and catagorizes colour
+        self.sensed_colour = self.colour_sensor.getColor()
 
-    def piston_down(self):
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kReverse)
+        distances = {}
+        for col_name, col_value in self.WHEEL_COLOURS.items():
+            distances[col_name] = col_value.dist(self.sensed_colour)
+
+        if (
+            distances[min(distances, key=lambda col: distances[col])]
+            < self.MAX_COLOUR_DIST
+        ):
+            return min(distances, key=lambda col: distances[col])
+
+    def get_wheel_dist(self) -> int:
+        # find the distance of the colour wheel from desired state in segments
+        # reuturns between -2 and 2 with 0 being the colour desired
+        current_colour = (
+            self.get_current_colour()
+        )  # only called when running position control
+        if current_colour == None:
+            return 0
+        distance = (
+            self.WHEEL_ORDER.index(self.required_colour)
+            + self.COLOUR_OFFSET
+            - self.WHEEL_ORDER.index(current_colour)
+        )
+        if distance > 2:
+            return distance - 4
+        return distance
+
+    def go_to_colour(self, colour: str) -> bool:  # do position control
+        self.required_colour = colour
+        distance = self.get_wheel_dist()
+        self.spinner_motor.set(distance * self.POSITION_SPIN_FACTOR)
```

Comment:
Calls like this that actually send commands to motors belong in execute. Other functions should be setting the parameters required for the motor calls. See [here](https://robotpy.readthedocs.io/en/stable/frameworks/magicbot.html#)